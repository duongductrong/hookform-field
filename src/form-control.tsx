"use client"

import { Slot } from "@radix-ui/react-slot"
import React, { forwardRef } from "react"
import { useFormField } from "./hooks/use-form-field"

export interface FormControlProps
  extends React.ComponentPropsWithoutRef<typeof Slot> {}

export const FormControl = forwardRef<
  React.ElementRef<typeof Slot>,
  FormControlProps
>(({ ...props }, ref) => {
  const { error, formItemId, formDescriptionId, formMessageId } = useFormField()

  return (
    <Slot
      ref={ref}
      id={formItemId}
      aria-describedby={
        !error
          ? `${formDescriptionId}`
          : `${formDescriptionId} ${formMessageId}`
      }
      aria-invalid={!!error}
      {...props}
    />
  )
})

FormControl.displayName = "FormControl"
