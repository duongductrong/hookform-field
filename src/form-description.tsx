"use client"

import { forwardRef, HTMLAttributes } from "react"
import { useFormField } from "./hooks/use-form-field"
import { ForwardRefComponent } from "./react-polymorphic"
import { cn } from "./utils"

export interface FormDescriptionProps
  extends HTMLAttributes<HTMLParagraphElement> {}

export const FormDescription = forwardRef(
  ({ component: Comp = "p", className, ...props }, ref) => {
    const { formDescriptionId } = useFormField()

    return (
      <Comp
        ref={ref}
        id={formDescriptionId}
        className={cn("field-description", className)}
        {...props}
      />
    )
  }
) as ForwardRefComponent<"p", FormDescriptionProps>

FormDescription.displayName = "FormDescription"
