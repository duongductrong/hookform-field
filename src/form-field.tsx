/* eslint-disable react/jsx-no-constructed-context-values */
import {
  Controller,
  ControllerProps,
  FieldPath,
  FieldValues,
} from "react-hook-form"
import FormFieldContext from "./context/form-field-context"

export const FormField = <
  TFieldValues extends FieldValues = FieldValues,
  TName extends FieldPath<TFieldValues> = FieldPath<TFieldValues>,
>({
  ...props
}: ControllerProps<TFieldValues, TName>) => (
  <FormFieldContext.Provider value={{ name: props.name }}>
    <Controller {...props} />
  </FormFieldContext.Provider>
)
