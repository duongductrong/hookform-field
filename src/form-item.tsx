"use client"

import { forwardRef, HTMLAttributes, useId, useMemo } from "react"
import FormItemContext from "./context/form-item-context"
import { cn } from "./utils"

export interface FormItemProps extends HTMLAttributes<HTMLDivElement> {}

export const FormItem = forwardRef<HTMLDivElement, FormItemProps>(
  ({ className, ...props }, ref) => {
    const id = useId()

    return (
      <FormItemContext.Provider value={useMemo(() => ({ id }), [id])}>
        <div ref={ref} className={cn("field-item", className)} {...props} />
      </FormItemContext.Provider>
    )
  }
)

FormItem.displayName = "FormItem"
