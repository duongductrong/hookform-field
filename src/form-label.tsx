"use client"

import React, { forwardRef } from "react"
import { useFormField } from "./hooks/use-form-field"
import { cn } from "./utils"

export interface FormLabelProps
  extends React.ComponentPropsWithoutRef<"label"> {}

export const FormLabel = forwardRef<React.ElementRef<"label">, FormLabelProps>(
  ({ className, ...props }, ref) => {
    const { error, formItemId, name } = useFormField()

    return (
      <label
        ref={ref}
        className={cn("field-label", className)}
        data-name={name}
        data-state={error ? "error" : "idle"}
        htmlFor={formItemId}
        {...props}
      />
    )
  }
)

FormLabel.displayName = "FormLabel"
