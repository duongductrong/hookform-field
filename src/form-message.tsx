"use client"

import React from "react"
import { useFormField } from "./hooks/use-form-field"
import { ForwardRefComponent } from "./react-polymorphic"
import { cn } from "./utils"

export interface FormMessageProps
  extends React.HTMLAttributes<HTMLParagraphElement> {}

export const FormMessage = React.forwardRef(
  ({ component: Comp = "p", className, children, ...props }, ref) => {
    const { name, error, formMessageId } = useFormField()
    const body = error ? String(error?.message) : children

    if (!body) {
      return null
    }

    return (
      <Comp
        ref={ref}
        id={formMessageId}
        className={cn("field-message", className)}
        data-state={error ? "error" : "idle"}
        data-name={name}
        {...props}
      >
        {body}
      </Comp>
    )
  }
) as ForwardRefComponent<"p", FormMessageProps>

FormMessage.displayName = "FormMessage"
