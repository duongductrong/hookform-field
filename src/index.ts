export * from "./field"
export * from "./form-control"
export * from "./form-description"
export * from "./form-field"
export * from "./form-item"
export * from "./form-label"
export * from "./form-message"
export * from "./hooks/use-form-field"
export * from "./utils"
export * from "./form"
