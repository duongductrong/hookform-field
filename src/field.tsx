"use client"

import { HTMLInputTypeAttribute, ReactNode, Suspense, memo } from "react"
import { get } from "react-hook-form"
import { FormControl } from "./form-control"
import { FormDescription } from "./form-description"
import { FormField } from "./form-field"
import { FormItem } from "./form-item"
import { FormLabel } from "./form-label"
import { FormMessage } from "./form-message"
import { cn } from "./utils"

export interface CreateFieldOptions {
  defineClassNames: {
    description?: string
    label?: string
    message?: string
    input?: string
    root?: string
  }
}

type InferComponentProps<T> = T extends React.ComponentType<infer P> ? P : never

export type FieldBaseProps = {
  name: string

  className?: string
  inputClassName?: string
  labelClassName?: string
  messageClassName?: string
  descriptionClassName?: string
  ClassName?: string

  description?: string
  disabled?: boolean
  id?: string
  label?: string
  placeholder?: string
  required?: boolean
  type?: HTMLInputTypeAttribute

  suspenseFallback?: ReactNode
}

export function createField<T extends Record<string, React.ComponentType<any>>>(
  components: T,
  options?: CreateFieldOptions
) {
  type ComponentName = keyof T
  type SyntheticFieldProps = { component: ComponentName } & FieldBaseProps &
    {
      [K in keyof T]: { component: K } & InferComponentProps<T[K]>
    }[ComponentName]

  const Field: React.FC<SyntheticFieldProps> = ({
    name,
    component,
    className,
    inputClassName,
    labelClassName,
    messageClassName,
    descriptionClassName,
    description,
    label,
    suspenseFallback,
    ...baseProps
  }) => {
    const customClassnames = options?.defineClassNames
    const InputComp = components[component]

    return (
      <Suspense fallback={suspenseFallback}>
        <FormField
          name={name}
          render={({ field, formState: { errors } }) => {
            const errorState = get(errors, name)
            return (
              <FormItem className={cn(customClassnames?.root, className)}>
                {label ? (
                  <FormLabel
                    className={cn(customClassnames?.label, labelClassName)}
                  >
                    {label}
                    {/* TODO: Will implement soon. */}
                    {/* {baseProps?.required ? (
                      <AsteriskIcon className="ml-1 h-3 w-3" />
                    ) : null} */}
                  </FormLabel>
                ) : null}
                <FormControl>
                  <InputComp
                    {...field}
                    {...baseProps}
                    className={cn(
                      inputClassName,
                      customClassnames?.input,
                      errorState?.message ? "error" : null
                    )}
                    data-state={errorState ? "error" : "idle"}
                  />
                </FormControl>

                {errorState?.message ? null : description ? (
                  <FormDescription
                    className={cn(
                      customClassnames?.description,
                      descriptionClassName
                    )}
                  />
                ) : null}

                {errorState && errorState.message ? (
                  <FormMessage
                    className={cn(customClassnames?.message, messageClassName)}
                  />
                ) : null}
              </FormItem>
            )
          }}
        />
      </Suspense>
    )
  }

  return memo(Field)
}
