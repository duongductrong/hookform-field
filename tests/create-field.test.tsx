import renderer from "react-test-renderer"
import { expect, it } from "vitest"
import { Form, createField } from "../src"

interface InputProps {}

interface NumberProps {}

interface FileProps {}

interface SelectProps {
  options: any[]
}

it("renders correctly", () => {
  const Field = createField({
    text: (props: InputProps) => <input type="text" />,
    number: (props: NumberProps) => <input type="number" />,
    file: (props: FileProps) => <input type="file" />,
    select: (props: SelectProps) => (
      <select>
        <option value="test">Test</option>
      </select>
    ),
  })

  const MyForm = () => {
    return (
      <Form>
        <Field label="Input" component="text" name="a" />
        <Field component="number" name="b" />
        <Field label="File" component="file" name="c" />
        <Field component="select" name="d" options={[]} />
      </Form>
    )
  }
  const form = renderer.create(<MyForm />)
  const formInstance = form.root

  expect(formInstance).toBeDefined()

  expect(
    formInstance.findByProps({ component: "text", name: "a" })
  ).toBeDefined()

  expect(
    formInstance.findByProps({ component: "number", name: "b" })
  ).toBeDefined()

  expect(
    formInstance.findByProps({ component: "file", name: "c" })
  ).toBeDefined()

  expect(
    formInstance.findByProps({ component: "select", name: "d" })
  ).toBeDefined()

  expect(form.toJSON()).toMatchSnapshot()
})

it("renders correctly custom components", () => {
  const Field = createField(
    {
      text: (props: InputProps) => <input type="text" />,
      number: (props: NumberProps) => <input type="number" />,
      file: (props: FileProps) => <input type="file" />,
      select: (props: SelectProps) => (
        <select>
          <option value="test">Test</option>
        </select>
      ),
    },
    {
      components: {
        root: "section",
        label: "p",
        description: "small",
        message: "small",
      },
    }
  )

  const MyForm = () => {
    return (
      <Form>
        <Field label="Input" component="text" name="a" />
        <Field component="number" name="b" />
        <Field label="File" component="file" name="c" />
        <Field component="select" name="d" options={[]} />
      </Form>
    )
  }

  const form = renderer.create(<MyForm />)

  expect(form.toJSON()).toMatchSnapshot()
})
