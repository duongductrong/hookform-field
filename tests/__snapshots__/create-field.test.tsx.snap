// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`renders correctly 1`] = `
<form>
  <div
    className="field-item"
  >
    <label
      className="field-label"
      data-name="a"
      data-state="idle"
      htmlFor=":r0:-form-item"
    >
      Input
    </label>
    <input
      type="text"
    />
  </div>
  <div
    className="field-item"
  >
    <input
      type="number"
    />
  </div>
  <div
    className="field-item"
  >
    <label
      className="field-label"
      data-name="c"
      data-state="idle"
      htmlFor=":r2:-form-item"
    >
      File
    </label>
    <input
      type="file"
    />
  </div>
  <div
    className="field-item"
  >
    <select>
      <option
        value="test"
      >
        Test
      </option>
    </select>
  </div>
</form>
`;

exports[`renders correctly custom components 1`] = `
<form>
  <section
    className="field-item"
  >
    <p
      className="field-label"
      data-name="a"
      data-state="idle"
      htmlFor=":r4:-form-item"
    >
      Input
    </p>
    <input
      type="text"
    />
  </section>
  <section
    className="field-item"
  >
    <input
      type="number"
    />
  </section>
  <section
    className="field-item"
  >
    <p
      className="field-label"
      data-name="c"
      data-state="idle"
      htmlFor=":r6:-form-item"
    >
      File
    </p>
    <input
      type="file"
    />
  </section>
  <section
    className="field-item"
  >
    <select>
      <option
        value="test"
      >
        Test
      </option>
    </select>
  </section>
</form>
`;
